File: the_cup.cpp
#include<iostream>using namespace std;class Cup {  public:    Cup       ();    void Fill ();    void Spill();  private:    bool filled_;};Cup::Cup () {  filled_ = false;}void Cup::Fill() {  if (filled_) {    cout << "The cup was already filled. Water is spilling everywhere.\n";  } else {    filled_ = true;    cout << "You have filled the cup with water. It looks slightly hazy.\n";  }}void Cup::Spill() {  if (filled_) {    filled_ = false;    cout << "You have purposely spilled the water from the cup. ";    cout << "You are a menace to society.\n";  } else {    cout << "You attempt to spill water from an empty cup. ";    cout << "Your peers are concerned about your mental health.\n";  }}int main() {  int i;  bool done = false;  Cup cup;  cout << "You have an empty cup.\n";  cout << "1. Fill it\n";  cout << "2. Spill it\n";  cout << "3. Move on.\n";  cin >> i;  do {    switch (i) {      case 1:        cup.Fill();        break;      case 2:        cup.Spill();        break;      case 3:        done = true;        break;      default:        cout << "Nothing you just did was related to the cup.\n";        break;    }    if (done)      break;  } while (cin >> i);  cout << "The cup won't miss you.\n";  return 0;}

File: the_cup.rb
class Cup
  attr_accessor :filled

  def initialize
    filled = false
  end

  def fill
    if filled
      puts "The cup was already filled. Water is spilling everywhere.\n"
    else
      puts "You have filled the cup with water. It looks slightly hazy.\n"
      self.filled = true
    end
  end

  def spill
    if filled
      puts "You have purposely spilled the water from the cup. "\
           "You are a menace to society.\n"
      self.filled = false
    else
      puts "You attempt to spill water from the empty cup. "\
           "Your peers are concerned about your mental health.\n"
    end
  end

end

cup = Cup.new
puts "You have an empty cup."
puts "1. Fill it\n"\
     "2. Spill it\n"\
     "3. Move on.\n"

until (input = gets.chomp) == '3'
  if input == '1'
    cup.fill
  elsif input == '2'
    cup.spill
  else
    puts "Nothing you just did was related to the cup"
  end
end
puts "The cup won't miss you."
